### ISSUE TITLE
Add derivative tests for layers

### ISSUE THREAD

**Issue by dan-zheng** [tensorflow/swift-apis#402]
Layers currently lack derivative tests.

* Tensor operation derivatives are tested in `Tests/TensorFlowTests/TensorAutoDiffTests.swift`.
* Layers are tested in `Tests/TensorFlowTests/LayerTests.swift`. (#77)

Consider adding layer derivative tests to `Tests/TensorFlowTests/LayerTests.swift`, alongside layer "forward-pass" tests.

**Comment by Shashi456** [comment]
@dan-zheng might be the right time to make a LayerTests directory and have forwardpass and differentiation tests as files?

**Comment by jkrukowski** [comment]
@Shashi456 @dan-zheng just wanted to ask you, what's the preferred way of implementing derivative tests for layers. Here is one of the first attempts https://github.com/tensorflow/swift-apis/pull/443 I'm thinking to use `pullback` function for that, something like
```swift
// ...
let seed = Tensor<Float>([[1, 1]])
let backprop = pullback(at: input, layer) { $1($0) }
let (_, 𝛁layer) = backprop(seed)
// ...
```

**Comment by Shashi456** [comment]
The preferred way is at it is done for batch normalization in `tensorAutoDiffTests.swift`. You can check that [here](https://github.com/tensorflow/swift-apis/blob/bc047be5c55d2125f4115d80dd76811163c9eb00/Tests/TensorFlowTests/TensorAutoDiffTests.swift#L409).

**Comment by jon-tow** [comment]
**Layer Gradient Tests**

- [ ] Sequential
- [ ] Conv1D
- [ ] Conv2D
- [ ] Conv3D
- [ ] DepthConv2D
- [ ] SeparableConv2D
- [ ] ZeroPadding1D
- [ ] ZeroPadding2D
- [ ] ZeroPadding3D
- [x] MaxPool1D 
- [x] MaxPool2D
- [x] MaxPool3D
- [ ] AvgPool1D
- [ ] AvgPool2D
- [ ] AvgPool3D
- [ ] GlobalAvgPool1D
- [ ] GlobalAvgPool2D
- [ ] GlobalAvgPool3D
- [ ] GlobalMaxPool1D
- [ ] GlobalMaxPool2D
- [ ] GlobalMaxPool3D
- [ ] UpSampling1D
- [ ] UpSampling2D
- [ ] UpSampling3D
- [ ] Reshape
- [ ] Flatten
- [ ] Embedding
- [x] Dense
- [ ] RNN
- [ ] SimpleRNNCell
- [ ] LSTM
- [ ] Function
- [x] BatchNorm
- [x] LayerNorm

**Comment by rxwei** [comment]
Thanks for the checklist @jon-tow !

**Comment by Shashi456** [comment]
#443 #478 #470 #460 #469

**Comment by sainikhit2k** [comment]
Is this list updated ?
